lib/QueryValidationVisitor.js

```diff
@@ -0,0 +1,229 @@
+    const argName = arg.name.value
+    // look for directive and validate argument if directive found
+    const argTypeDef = this.currentrFieldDef.args.find(d => d.name === argName)
+    const value = valueFromAST(arg.value, argTypeDef.type, this.variableValues)
+
+    // nothing to validate
+    if (!value) return
+
+    let variableName
+    if (arg.value.kind === Kind.VARIABLE) variableName = arg.value.name.value
+
+    let valueTypeDef = argTypeDef.type
+    if (isNonNullType(valueTypeDef)) valueTypeDef = valueTypeDef.ofType
+
+    if (isInputObjectType(valueTypeDef)) {
+      const inputObjectTypeDef = getNamedType(valueTypeDef)
+      validateInputTypeValue(this.context, inputObjectTypeDef, argName, variableName, value, this.currentField, variableName)
+    } else if (isListType(valueTypeDef)) {
+      validateArrayTypeValue(this.context, valueTypeDef, argTypeDef, value, this.currentField, argName, variableName, variableName)
+    } else {
+      const fieldNameForError = variableName || this.currentField.name.value + '.' + argName
+      validateScalarTypeValue(this.context, this.currentField, argTypeDef, valueTypeDef, value, variableName, argName, fieldNameForError, '')
+    }
+  }
+}
+
+function validateScalarTypeValue (context, currentQueryField, typeDefWithDirective, valueTypeDef, value, variableName, argName, fieldNameForError, errMessageAt) {
+  const directiveArgumentMap = getDirectiveValues(constraintDirectiveTypeDefsObj, typeDefWithDirective.astNode)
+  if (directiveArgumentMap) {
+    const st = getScalarType(valueTypeDef).scalarType
+    const valueDelim = st === GraphQLString ? '"' : ''
+    try {
+      getConstraintValidateFn(st)(fieldNameForError, directiveArgumentMap, value)
+    } catch (e) {
+      let error
+      if (variableName) {
+        error = new ValidationError(fieldNameForError, `Variable "$${variableName}" got invalid value ${valueDelim}${value}${valueDelim}${errMessageAt}. ` + e.message, e.context)
+      } else {
+        error = new ValidationError(fieldNameForError, `Argument "${argName}" of "${currentQueryField.name.value}" got invalid value ${valueDelim}${value}${valueDelim}${errMessageAt}. ` + e.message, e.context)
+      }
+      error.originalError = e
+      context.reportError(error)
+    }
+  }
+}
+
+function validateInputTypeValue (context, inputObjectTypeDef, argName, variableName, value, currentField, parentNames) {
+  // use new visitor to traverse input object structure
+  const visitor = new InputObjectValidationVisitor(context, inputObjectTypeDef, argName, variableName, value, currentField, parentNames)
+  visit(inputObjectTypeDef.astNode, visitor)
+}
+
+function validateArrayTypeValue (context, valueTypeDef, typeDefWithDirective, value, currentField, argName, variableName, iFieldNameFull) {
+  let valueTypeDefArray = valueTypeDef.ofType
+  if (isNonNullType(valueTypeDefArray)) valueTypeDefArray = valueTypeDefArray.ofType
+
+  let i = 0
+  value.forEach(element => {
```

Comment:
Can we not use the `index` from the `forEach` instead of `i`?

Reply:
Ah, sure, man is constantly learning :-D